Partially fix `explicit_outlives_requirements` lint in macros

lukas-code opened the pull request:
Show the suggestion if and only if the bounds are from the same source context.

fixes https://github.com/rust-lang/rust/issues/106044
fixes https://github.com/rust-lang/rust/issues/106063

Commit:
Fix `explicit_outlives_requirements` lint in macros

Show the suggestion if and only if the bounds are from the same source context.

lcnr commented:
r? compiler (on vacation)

cjgillot commented on compiler/rustc_ast_lowering/src/item.rs:
```diff
@@ -1377,21 +1377,8 @@ impl<'hir> LoweringContext<'_, 'hir> {
-            })
-            .unwrap_or(param_span.shrink_to_hi());
+        let span =
+            bounds.iter().fold(param_span.shrink_to_hi(), |span, bound| span.to(bound.span()));
```
IMO, the issue is that we are trying to reconstruct a span during lowering, which will mix spans coming from macro different expansions.
Fortunately, we know what `SyntaxContext` produces the `:`, it's `GenericParam::colon_span`.
So there are 2 cases here:
- if `colon_span` is `None`, we should use `param_span.shrink_to_hi()`;
- if not, we should use it, and extend it with `bound.span().find_ancestor_in_same_ctxt(colon_span.ctxt())`.

lukas-code replied:
`bound.span().find_ancestor_in_same_ctxt(colon_span.ctxt())` goes horribly wrong when the colon also comes from a macro parameter:

https://github.com/rust-lang/rust/blob/d379a25e415e1291ced6c8bbf03c080d638ecc49/src/test/ui/rust-2018/edition-lint-infer-outlives-macro.fixed#L89-L96

I changed this to search only in the "parent span", which is either the entire generics list for (normal) generic args or the `impl Trait` span.

cjgillot commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -2184,30 +2184,36 @@ impl ExplicitOutlivesRequirements {
+                    return None;
+                }
+
+                let span = bound.span();
```
```suggestion
                let span = bound.span().find_ancestor_inside(whole_predicate_span)?
```
This will allow to support a few more cases.

Commit:
document that `Span::to` can go backwards

Commit:
address review comments + better tests

lukas-code commented:
I've addressed the review comments and this PR also fixes #106063 now. There are, however, still some remaining inconsistencies. See the `FIXME` at the bottom of the UI test for an example.

This happens, because the where clause span gets messed up in the parser here:
https://github.com/rust-lang/rust/blob/a1fc71196a5554eafc5dd9a1b4e0c159717141e0/compiler/rustc_parse/src/parser/generics.rs#L248-L315

... due to incomplete handling of mixed macro expansions in `Span::to`:
https://github.com/rust-lang/rust/blob/a1fc71196a5554eafc5dd9a1b4e0c159717141e0/compiler/rustc_span/src/lib.rs#L810-L818

This is presumably a problem elsewhere too and should probably be addressed in general (in a different PR) instead of adding more special casing here.

@rustbot review

cjgillot commented:
I'm not sure I understand what is the issue with your FIXME.
Anyway, this is a great improvement!
@bors r+

bors commented:
:pushpin: Commit 1eba6c404fa3e64607fc3826214bd61ae10fa5bb has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 1eba6c404fa3e64607fc3826214bd61ae10fa5bb -->
<!-- homu: {"type":"Approved","sha":"1eba6c404fa3e64607fc3826214bd61ae10fa5bb","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
